sui_programmability/examples/frenemies/sources/registry.move

```diff
@@ -0,0 +1,140 @@
+module frenemies::registry {
+    use std::string::{Self, String};
+    use std::option::{Self, Option};
+    use sui::object::{Self, UID};
+    use sui::table::{Self, Table};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    friend frenemies::frenemies;
+
+    /// Shared registry ensuring global uniqueness of names + tracking total participants
+    struct Registry has key {
+        id: UID,
+        players: Table<Name, address>,
+    }
+
+    /// Wrapper around String that enforces a max name length
+    struct Name has copy, drop, store {
+        name: String
+    }
+
+    /// This name exceeds the max length
+    const ENameTooLong: u64 = 0;
+
+    /// This name has already been registered by a different player
+    const ENameAlreadyRegistered: u64 = 1;
+
+    /// The maximum length of a username
+    const MAX_NAME_SIZE: u64 = 64;
+
+    fun init(ctx: &mut TxContext) {
+        transfer::share_object(Registry {
+            id: object::new(ctx),
+            players: table::new(ctx),
+        })
+    }
+
+    /// Only callable from scorecard
+    public(friend) fun register(self: &mut Registry, name: String, ctx: &TxContext): Name {
+        assert!(string::length(&name) <= MAX_NAME_SIZE, ENameTooLong);
+
+        let name = Name { name };
+        let players = &mut self.players;
+        assert!(!table::contains(players, name), ENameAlreadyRegistered);
```

Comment:
With this setup, an address can register again but with a different name. Is this something we want to allow?

Reply:
Oh from the tests it seems that we are fine with it. Why though?

Reply:
There's no harm in registering multiple times with the same address, and it doesn't give a player any advantages because all the scorecards belonging to the same address will get the same assignment. We could prevent this with an extra table, but I thought I'd allow it so folks can explore the strategy of getting multiple scorecards and then discover that it doesn't help :)